src/jit/emitxarch.cpp

```diff
@@ -2901,174 +2901,283 @@ void emitter::emitInsStoreLcl(instruction ins, emitAttr attr, GenTreeLclVarCommo
+                        {
+                            assert(cnsOp == src);
+                            assert(otrOp == nullptr);
+                            assert(src->IsCnsIntOrI());
+
+                            fmt = emitInsModeFormat(ins, IF_ARD_CNS);
+                        }
+                        else
+                        {
+                            assert(otrOp == src);
+                            fmt = emitInsModeFormat(ins, IF_ARD_RRD);
+                        }
+                    }
+                    assert(fmt != IF_NONE);
+                    emitHandleMemOp(memIndir, id, fmt, ins);
+
+                    // Determine the instruction size
+                    UNATIVE_OFFSET sz = 0;
+
+                    if (memOp == src)
+                    {
+                        assert(otrOp == dst);
+                        assert(cnsOp == nullptr);
 
-        if (memBase->OperGet() == GT_LCL_VAR_ADDR)
+                        if (instrHasImplicitRegPairDest(ins))
+                        {
+                            sz = emitInsSizeAM(id, insCode(ins));
+                        }
+                        else
+                        {
+                            sz = emitInsSizeAM(id, insCodeRM(ins));
+                        }
+                    }
+                    else
+                    {
+                        assert(memOp == dst);
+
+                        if (cnsOp != nullptr)
+                        {
+                            assert(memOp == dst);
+                            assert(cnsOp == src);
+                            assert(otrOp == nullptr);
+
+                            sz = emitInsSizeAM(id, insCodeMI(ins), (int)src->gtIntConCommon.IconValue());
+                        }
+                        else
+                        {
+                            assert(otrOp == src);
+                            sz = emitInsSizeAM(id, insCodeMR(ins));
+                        }
+                    }
+                    assert(sz != 0);
+
+                    id->idCodeSize(sz);
+
+                    dispIns(id);
+                    emitCurIGsize += sz;
+
+                    return (memOp == src) ? dst->gtRegNum : REG_NA;
+                }
+            }
+        }
+        else
         {
-            varNum = memBase->AsLclVarCommon()->GetLclNum();
-            offset = 0;
+            switch (memOp->OperGet())
+            {
+                case GT_LCL_FLD:
+                case GT_STORE_LCL_FLD:
+                {
+                    GenTreeLclFld* lclField = memOp->AsLclFld();
+                    varNum                  = lclField->GetLclNum();
+                    offset                  = lclField->gtLclFld.gtLclOffs;
+                    break;
+                }
+
+                case GT_LCL_VAR:
+                {
+                    assert(memOp->IsRegOptional() || !emitComp->lvaTable[memOp->gtLclVar.gtLclNum].lvIsRegCandidate());
```

Comment:
It would be better to split this assert in two. The second condition should be moved after the assignment below so it can reuse `varNum`.

Reply:
It's an `||` condition, so it can't be split... No matter how you do the `memOp->IsRegOptional()` check, it can fail on its own.

Reply:
Unless I'm just too tired to remember how to do this 😄

Reply:
Oops, missed the `||`.